recipe/meta.yaml

```diff
@@ -15,7 +16,8 @@ source:
     - python3interp.patch
 
 build:
-  number: 1008
+  number: {{ build }}
+  string: {{ mpi }}_{{ c_compiler }}_{{ build }}
```

Comment:
@minrk should I really update this?
How does a user decide which package to install later on with `conda install`? How is the variant encoded?

Reply:
I wonder why the variants are not show in `README.md` after rerender.

Reply:
I wrote up what I've found so far with the mpi variants [here](https://hackmd.io/ry4uI0thTs2q_b4mAQd_qg). The gist of the current strategy for packages with a `nompi` variant is:

1. prefer nompi over mpi by default with a build number offset of 100 (like the 1000 offset we have for new compilers)
2. *add* mpi variant to build string, preserving the package hash and build number
3. run_exports for *at least* the mpi variants, possibly all, depending on the given library's compatibility

> should I really update this? How is the variant encoded?

Thanks to package hashes, you only need to add to a build string what you want users or downstream packages to be able to pick explicitly. This is usually the case when you have a nompi variant, but not if it's only mpich vs openmpi. There, the package hash is enough and `conda install pkg mpich` will pick the mpich variant.

I've used [this build string](https://github.com/conda-forge/hdf5-feedstock/blob/ad3ef4eee81f6f4c7dff9e09cd7d97655dafffb1/recipe/meta.yaml#L38-L62). It includes `PKG_HASH` which ensures that compiler variants etc. are still properly captured.

> How does a user decide which package to install later on with conda install?

The above build string and build number offset creates the following builds, assuming you are still building with two compiler stacks:

    # new compilers
    adios-1.13.1-nompi_habc123_1109
    adios-1.13.1-mpi_mpich_habc123_1009
    adios-1.13.1-mpi_openmpi_habc123_1009
    # old compilers
    adios-1.13.1-nompi_habc456_109
    adios-1.13.1-mpi_mpich_habc456_9
    adios-1.13.1-mpi_openmpi_habc456_9


which allows users to make the following installation choices:

- `conda install adios=*=mpi*` installs any mpi-compatible implementation
-  `conda install adios=*=mpi_mpich*` explicitly picks mpich variant. This will have the same result as `conda install mpich adios=*=mpi*`
- `conda install adios=1.13=nompi*` explicitly forbids mpi. Usually not necessary, as the build number offset will cause nompi to be preferred unless mpi is explicitly requested.

> I wonder why the variants are not show in README.md after rerender.

Since they are all variants the same package, I don't think they should show up in the README. However, another strategy for mpi variants is to create a different package *name* for the mpi builds, as is done [in mumps](https://github.com/conda-forge/mumps-feedstock/blob/master/recipe/meta.yaml) or [scotch](https://github.com/conda-forge/scotch-feedstock/blob/master/recipe/meta.yaml). Which strategy to choose will depend on how your library works (i.e. there should not be file conflicts between the mpi and non-mpi variants). However, if it does make sense for your package to build this way, it can be simpler since you don't need to do anything with the build string.

Reply:
Thanks a ton, that helps a lot!
Just one question on the package string: first, great with `PKG_HASH`, that's what I was concerned about. But second, why do we at all add `nompi_` in the package string? We could, as in other cases, just skip it and only add one for the two other builds with mpi flavors, no? Probably that's just a style question.

Reply:
It is mostly a style question, but it does make it easier for users to explicitly block mpi variants if/when that's desirable rather than relying on the solver priority, which would be harder to specify otherwise.